extensions/amp-ad/0.1/amp-ad.js

```diff
@@ -54,44 +55,48 @@ function copyAttributes(sourceElement, targetElement) {
 }
 
 export class AmpAd extends AMP.BaseElement {
-  constructor(element) {
-    super(element);
-  }
-
-  /** @override */
-  isLayoutSupported(layout) {
-    return isLayoutSizeDefined(layout);
-  }
 
   /** @override */
-  buildCallback() {
+  upgradeCallback() {
     const type = this.element.getAttribute('type');
     if (!type) {
       // Unspecified or empty type.  Nothing to do here except bail out.
-      return;
+      return null;
     }
-    let newChild;
     // TODO(tdrl): Check amp-ad registry to see if they have this already.
     if (!a4aRegistry[type] ||
         !a4aRegistry[type](this.getWin(), this.element)) {
       // Network either has not provided any A4A implementation or the
       // implementation exists, but has explicitly chosen not to handle this
       // tag as A4A.  Fall back to the 3p implementation.
-      newChild = this.element.ownerDocument.createElement(TAG_3P_IMPL);
-    } else {
-      // Note: The insertAmpExtensionScript method will pick the version number.
-      // If we ever reach a point at which there are different extensions with
-      // different version numbers at play simultaneously, we'll have to make sure
-      // that the loader can handle the case.
-      const extensionTag = networkImplementationTag(type);
-      newChild = this.element.ownerDocument.createElement(extensionTag);
-      /*OK*/insertAmpExtensionScript(this.getWin(), extensionTag, true);
+      return new AmpAd3PImpl(this.element);
```

Comment:
Oh, right, and one more:

- I assume this will handle replaying any events that arrived at the `<amp-ad>` element before it is upgraded to the post-upgrade element?